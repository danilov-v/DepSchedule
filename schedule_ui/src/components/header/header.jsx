import React, { useState } from "react";
import {
  Collapse,
  Navbar,
  NavbarToggler,
  Nav,
  NavItem,
  NavLink,
  UncontrolledDropdown,
  DropdownToggle,
  DropdownMenu,
  DropdownItem,
} from "reactstrap";
import { useSelector, useDispatch } from "react-redux";
import { Link, useLocation } from "react-router-dom";
import classnames from "classnames";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import DatePicker from "react-datepicker";
import { updateCalendar } from "redux/actions/calendars";
import { logoutRequest } from "redux/actions/auth";
import { getActiveCalendarInfo } from "redux/selectors/calendars";
import { getUserDataSelector } from "redux/selectors/ui";
import { setActiveCalendar } from "redux/actions/calendars";
import { setStartDate, setEndDate } from "redux/actions/ui";
import logo from "logo.png";
import { LastEventsList } from "components/last-events-list/last-events-list";

import "./header.scss";

export function Header() {
  const dispatch = useDispatch();
  const location = useLocation();
  const { startDate, endDate, operationalDate } = useSelector(
    getActiveCalendarInfo
  );
  const { startDate: userStartDate, endDate: userEndDate } = useSelector(
    getUserDataSelector
  );

  const [isOpen, toggleNav] = useState(false);
  const isHomePage = location.pathname === "/";

  // Handlers
  const changeStartDate = date => dispatch(setStartDate(date));
  const changeEndDate = date => dispatch(setEndDate(date));
  const changeOperationalDate = date =>
    dispatch(updateCalendar({ operationalDate: date }));
  const changeCalendar = () =>
    dispatch(setActiveCalendar({ calendarId: null }));
  const logout = () => dispatch(logoutRequest());
  const toggle = () => toggleNav(!isOpen);
  const print = () => window.print();

  return (
    <Navbar
      color="light"
      light
      expand="lg"
      className="header mb-3 bg-transparent"
    >
      <Link className="nav-link brand" to="/">
        <img src={logo} alt="Logo" className="logo" />
        ГРСУ
      </Link>
      <NavbarToggler onClick={toggle} />
      <Collapse isOpen={isOpen} navbar>
        <Nav className="ml-auto" navbar>
          <div
            title="Оперативное время"
            className={classnames("d-flex align-items-center mr-5", {
              invisible: !isHomePage,
            })}
          >
            <span>Оперативное время: &nbsp;</span>
            <DatePicker
              selected={operationalDate}
              dateFormat="dd/MM/yyyy"
              onChange={changeOperationalDate}
              locale="ru"
              className="date_picker"
            />
          </div>
          <div
            className={classnames("d-flex align-items-center mr-5", {
              invisible: !isHomePage,
            })}
          >
            События с &nbsp;
            <DatePicker
              minDate={startDate}
              maxDate={endDate}
              selected={userStartDate}
              dateFormat="dd/MM/yyyy"
              onChange={changeStartDate}
              locale="ru"
              className="date_picker"
            />
            &nbsp; по &nbsp;
            <DatePicker
              selected={userEndDate}
              minDate={startDate}
              maxDate={endDate}
              dateFormat="dd/MM/yyyy"
              onChange={changeEndDate}
              locale="ru"
              className="date_picker"
            />
          </div>
          <NavItem title="Типы событий">
            <Link className="nav-link" to="event_types">
              Типы событий
            </Link>
          </NavItem>
          <NavItem title="Периоды">
            <Link className="nav-link" to="periods">
              Периоды
            </Link>
          </NavItem>
          <NavItem title="Последние события">
            <LastEventsList />
          </NavItem>
          <NavItem hidden={!isHomePage} title="Печать">
            <NavLink href="#" onClick={print}>
              <FontAwesomeIcon icon="file-pdf" />
            </NavLink>
          </NavItem>
          <UncontrolledDropdown nav inNavbar title="Управление">
            <DropdownToggle nav>
              <FontAwesomeIcon icon="user-alt" />
            </DropdownToggle>
            <DropdownMenu right>
              <DropdownItem onClick={changeCalendar}>
                Сменить календарь
              </DropdownItem>
              <DropdownItem divider />
              <DropdownItem onClick={logout}>Выйти</DropdownItem>
            </DropdownMenu>
          </UncontrolledDropdown>
        </Nav>
      </Collapse>
    </Navbar>
  );
}
