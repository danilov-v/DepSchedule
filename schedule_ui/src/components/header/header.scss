@import "variables.scss";

.header {
  position: sticky;
  top: 0;
  color: #00000080;
  z-index: 100;

  .brand {
    color: $color-grey-dark-3;
    font-size: 1.25em;
    font-weight: 600;
  }
  .logo {
    width: 55px;
    height: 30px;
    margin-right: 5px;
  }
  .date_picker {
    color: $color-grey-dark-3;
    border: 1px solid $color-grey-dark;
    border-radius: 5px;
  }
}
