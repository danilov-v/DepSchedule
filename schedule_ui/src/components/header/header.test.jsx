import React from "react";
import { shallow } from "enzyme";
import { set } from "lodash";
import { HeaderUI } from "./header";

const mockDate = new Date("2019-05-05");
const props = {
  startDate: mockDate,
  endDate: mockDate,
  operationalDate: mockDate,
  onChangeStartDate: () => {},
  onChangeEndDate: () => {},
  history: {
    location: {
      pathname: "/",
    },
  },
};

describe("header.jsx", () => {
  describe("<HeaderUI />", () => {
    test("should render and match snapshot for home page", () => {
      const wrapper = shallow(<HeaderUI {...props} />);

      expect(wrapper).toMatchSnapshot();
    });

    test("should render and match snapshot for NOT home page", () => {
      set(props, "history.location.pathname", "/test");

      const wrapper = shallow(<HeaderUI {...props} />);

      expect(wrapper).toMatchSnapshot();
    });
  });
});
