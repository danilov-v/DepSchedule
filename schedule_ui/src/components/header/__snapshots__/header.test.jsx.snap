// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header.jsx <HeaderUI /> should render and match snapshot for NOT home page 1`] = `
<Navbar
  className="header mb-3"
  color="light"
  expand="lg"
  light={true}
  tag="nav"
>
  <Link
    className="nav-link brand"
    to="/"
  >
    <img
      alt="Logo"
      className="logo"
      src="logo.png"
    />
    ГРСУ
  </Link>
  <NavbarToggler
    onClick={[Function]}
    tag="button"
    type="button"
  />
  <Collapse
    appear={false}
    enter={true}
    exit={true}
    in={false}
    isOpen={false}
    mountOnEnter={false}
    navbar={true}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    tag="div"
    timeout={350}
    unmountOnExit={false}
  >
    <Nav
      className="ml-auto"
      navbar={true}
      tag="ul"
      vertical={false}
    >
      <div
        className="d-flex align-items-center mr-5 invisible"
      >
        <span>
          Оперативное время:  
        </span>
        <DatePicker
          allowSameDay={false}
          className="date_picker"
          dateFormat="dd/MM/yyyy"
          dateFormatCalendar="LLLL yyyy"
          disabled={false}
          disabledKeyboardNavigation={false}
          dropdownMode="scroll"
          inlineFocusSelectedMonth={false}
          locale="ru"
          monthsShown={1}
          nextMonthButtonLabel="Next month"
          onBlur={[Function]}
          onChange={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onInputClick={[Function]}
          onInputError={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          onYearChange={[Function]}
          preventOpenOnFocus={false}
          previousMonthButtonLabel="Previous Month"
          readOnly={false}
          renderDayContents={[Function]}
          selected={2019-05-05T00:00:00.000Z}
          shouldCloseOnSelect={true}
          showMonthYearPicker={false}
          showTimeInput={false}
          showTimeSelect={false}
          strictParsing={false}
          timeCaption="Time"
          timeInputLabel="Time"
          timeIntervals={30}
          withPortal={false}
        />
      </div>
      <div
        className="d-flex align-items-center mr-5 invisible"
      >
        События с  
        <DatePicker
          allowSameDay={false}
          className="date_picker"
          dateFormat="dd/MM/yyyy"
          dateFormatCalendar="LLLL yyyy"
          disabled={false}
          disabledKeyboardNavigation={false}
          dropdownMode="scroll"
          inlineFocusSelectedMonth={false}
          locale="ru"
          monthsShown={1}
          nextMonthButtonLabel="Next month"
          onBlur={[Function]}
          onChange={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onInputClick={[Function]}
          onInputError={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          onYearChange={[Function]}
          preventOpenOnFocus={false}
          previousMonthButtonLabel="Previous Month"
          readOnly={false}
          renderDayContents={[Function]}
          selected={2019-05-05T00:00:00.000Z}
          shouldCloseOnSelect={true}
          showMonthYearPicker={false}
          showTimeInput={false}
          showTimeSelect={false}
          strictParsing={false}
          timeCaption="Time"
          timeInputLabel="Time"
          timeIntervals={30}
          withPortal={false}
        />
          по  
        <DatePicker
          allowSameDay={false}
          className="date_picker"
          dateFormat="dd/MM/yyyy"
          dateFormatCalendar="LLLL yyyy"
          disabled={false}
          disabledKeyboardNavigation={false}
          dropdownMode="scroll"
          inlineFocusSelectedMonth={false}
          locale="ru"
          minDate={2019-05-05T00:00:00.000Z}
          monthsShown={1}
          nextMonthButtonLabel="Next month"
          onBlur={[Function]}
          onChange={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onInputClick={[Function]}
          onInputError={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          onYearChange={[Function]}
          preventOpenOnFocus={false}
          previousMonthButtonLabel="Previous Month"
          readOnly={false}
          renderDayContents={[Function]}
          selected={2019-05-05T00:00:00.000Z}
          shouldCloseOnSelect={true}
          showMonthYearPicker={false}
          showTimeInput={false}
          showTimeSelect={false}
          strictParsing={false}
          timeCaption="Time"
          timeInputLabel="Time"
          timeIntervals={30}
          withPortal={false}
        />
      </div>
      <NavItem
        tag="li"
      >
        <Link
          className="nav-link"
          to="event_types"
        >
          Типы событий
        </Link>
      </NavItem>
      <NavItem
        tag="li"
      >
        <Link
          className="nav-link"
          to="periods"
        >
          Периоды
        </Link>
      </NavItem>
      <NavItem
        hidden={true}
        tag="li"
      >
        <NavLink
          href="#"
          onClick={[Function]}
          tag="a"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="file-pdf"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
        </NavLink>
      </NavItem>
      <UncontrolledDropdown
        inNavbar={true}
        nav={true}
      >
        <DropdownToggle
          aria-haspopup={true}
          color="secondary"
          nav={true}
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="user-alt"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
        </DropdownToggle>
        <DropdownMenu
          flip={true}
          right={true}
          tag="div"
        >
          <DropdownItem
            tag="button"
            toggle={true}
          >
            Профиль
          </DropdownItem>
          <DropdownItem
            divider={true}
            tag="button"
            toggle={true}
          />
          <DropdownItem
            tag="button"
            toggle={true}
          >
            Выйти
          </DropdownItem>
        </DropdownMenu>
      </UncontrolledDropdown>
    </Nav>
  </Collapse>
</Navbar>
`;

exports[`header.jsx <HeaderUI /> should render and match snapshot for home page 1`] = `
<Navbar
  className="header mb-3"
  color="light"
  expand="lg"
  light={true}
  tag="nav"
>
  <Link
    className="nav-link brand"
    to="/"
  >
    <img
      alt="Logo"
      className="logo"
      src="logo.png"
    />
    ГРСУ
  </Link>
  <NavbarToggler
    onClick={[Function]}
    tag="button"
    type="button"
  />
  <Collapse
    appear={false}
    enter={true}
    exit={true}
    in={false}
    isOpen={false}
    mountOnEnter={false}
    navbar={true}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    tag="div"
    timeout={350}
    unmountOnExit={false}
  >
    <Nav
      className="ml-auto"
      navbar={true}
      tag="ul"
      vertical={false}
    >
      <div
        className="d-flex align-items-center mr-5"
      >
        <span>
          Оперативное время:  
        </span>
        <DatePicker
          allowSameDay={false}
          className="date_picker"
          dateFormat="dd/MM/yyyy"
          dateFormatCalendar="LLLL yyyy"
          disabled={false}
          disabledKeyboardNavigation={false}
          dropdownMode="scroll"
          inlineFocusSelectedMonth={false}
          locale="ru"
          monthsShown={1}
          nextMonthButtonLabel="Next month"
          onBlur={[Function]}
          onChange={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onInputClick={[Function]}
          onInputError={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          onYearChange={[Function]}
          preventOpenOnFocus={false}
          previousMonthButtonLabel="Previous Month"
          readOnly={false}
          renderDayContents={[Function]}
          selected={2019-05-05T00:00:00.000Z}
          shouldCloseOnSelect={true}
          showMonthYearPicker={false}
          showTimeInput={false}
          showTimeSelect={false}
          strictParsing={false}
          timeCaption="Time"
          timeInputLabel="Time"
          timeIntervals={30}
          withPortal={false}
        />
      </div>
      <div
        className="d-flex align-items-center mr-5"
      >
        События с  
        <DatePicker
          allowSameDay={false}
          className="date_picker"
          dateFormat="dd/MM/yyyy"
          dateFormatCalendar="LLLL yyyy"
          disabled={false}
          disabledKeyboardNavigation={false}
          dropdownMode="scroll"
          inlineFocusSelectedMonth={false}
          locale="ru"
          monthsShown={1}
          nextMonthButtonLabel="Next month"
          onBlur={[Function]}
          onChange={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onInputClick={[Function]}
          onInputError={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          onYearChange={[Function]}
          preventOpenOnFocus={false}
          previousMonthButtonLabel="Previous Month"
          readOnly={false}
          renderDayContents={[Function]}
          selected={2019-05-05T00:00:00.000Z}
          shouldCloseOnSelect={true}
          showMonthYearPicker={false}
          showTimeInput={false}
          showTimeSelect={false}
          strictParsing={false}
          timeCaption="Time"
          timeInputLabel="Time"
          timeIntervals={30}
          withPortal={false}
        />
          по  
        <DatePicker
          allowSameDay={false}
          className="date_picker"
          dateFormat="dd/MM/yyyy"
          dateFormatCalendar="LLLL yyyy"
          disabled={false}
          disabledKeyboardNavigation={false}
          dropdownMode="scroll"
          inlineFocusSelectedMonth={false}
          locale="ru"
          minDate={2019-05-05T00:00:00.000Z}
          monthsShown={1}
          nextMonthButtonLabel="Next month"
          onBlur={[Function]}
          onChange={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onInputClick={[Function]}
          onInputError={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          onYearChange={[Function]}
          preventOpenOnFocus={false}
          previousMonthButtonLabel="Previous Month"
          readOnly={false}
          renderDayContents={[Function]}
          selected={2019-05-05T00:00:00.000Z}
          shouldCloseOnSelect={true}
          showMonthYearPicker={false}
          showTimeInput={false}
          showTimeSelect={false}
          strictParsing={false}
          timeCaption="Time"
          timeInputLabel="Time"
          timeIntervals={30}
          withPortal={false}
        />
      </div>
      <NavItem
        tag="li"
      >
        <Link
          className="nav-link"
          to="event_types"
        >
          Типы событий
        </Link>
      </NavItem>
      <NavItem
        tag="li"
      >
        <Link
          className="nav-link"
          to="periods"
        >
          Периоды
        </Link>
      </NavItem>
      <NavItem
        hidden={false}
        tag="li"
      >
        <NavLink
          href="#"
          onClick={[Function]}
          tag="a"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="file-pdf"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
        </NavLink>
      </NavItem>
      <UncontrolledDropdown
        inNavbar={true}
        nav={true}
      >
        <DropdownToggle
          aria-haspopup={true}
          color="secondary"
          nav={true}
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="user-alt"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
        </DropdownToggle>
        <DropdownMenu
          flip={true}
          right={true}
          tag="div"
        >
          <DropdownItem
            tag="button"
            toggle={true}
          >
            Профиль
          </DropdownItem>
          <DropdownItem
            divider={true}
            tag="button"
            toggle={true}
          />
          <DropdownItem
            tag="button"
            toggle={true}
          >
            Выйти
          </DropdownItem>
        </DropdownMenu>
      </UncontrolledDropdown>
    </Nav>
  </Collapse>
</Navbar>
`;
